modules/fedimint-wallet-client/src/lib.rs

```diff
@@ -215,6 +223,132 @@ impl ClientModuleInit for WalletClientInit {
             task_group: args.task_group().clone(),
         })
     }
+
+    /// Wallet recovery
+    ///
+    /// Query bitcoin rpc for history of addresses from last known used
+    /// addresses (or index 0 one) until MAX_GAP unused ones.
+    ///
+    /// Notably does not persist the progress of addresses being querried,
+    /// because it is not expected that it would take long enough to bother.
+    async fn recover(
+        &self,
+        args: &ClientModuleRecoverArgs<Self>,
+        _snapshot: Option<&<Self::Module as ClientModule>::Backup>,
+    ) -> anyhow::Result<()> {
+        let db = args.db().clone();
+        if db
+            .begin_transaction_nc()
+            .await
+            .get_value(&RecoveryFinalizedKey)
+            .await
+            .unwrap_or_default()
+        {
+            debug!(target: LOG_CLIENT_MODULE_WALLET, max_gap=MAX_GAP, "Recovery already complete before");
+            return Ok(());
+        }
+
+        let start_tweak_idx = match _snapshot {
+            Some(WalletModuleBackup::V0(backup)) => {
+                debug!(target: LOG_CLIENT_MODULE_WALLET, "Restoring starting from an existing backup");
+
+                backup.next_tweak_idx
+            }
+            _ => TweakIdx(0),
+        };
+
+        /// We will check this many addresses ahead since last actually used
+        /// one before we give up
+        const MAX_GAP: u64 = 10;
+        /// New client will start deriving new addresses from last_used one
+        /// plus that many idexes. This should be less than
+        /// MAX_GAP, but more than 0: We want to make sure we detect
+        /// deposits that might have been made after multiple re-derivation,
+        /// but we want also to avoid accidental address re-use.
+        const NUM_IDX_SKIP_AFTER_LAST_USED: u64 = 3;
+
+        let data = WalletClientModuleData {
+            cfg: args.cfg().clone(),
+            module_root_secret: args.module_root_secret().clone(),
+        };
+
+        let rpc_config = self
+            .0
+            .clone()
+            .unwrap_or(WalletClientModule::get_rpc_config(args.cfg()));
+
+        let btc_rpc = create_bitcoind(&rpc_config, TaskGroup::new().make_handle())?;
+
+        let mut last_used_idx = None;
+        let mut tweak_idx = start_tweak_idx;
+        let mut tweak_idxes_with_pegins = BTreeSet::new();
+
+        debug!(target: LOG_CLIENT_MODULE_WALLET, max_gap=MAX_GAP, %start_tweak_idx, "Checking history");
+
+        let new_start_tweak_idx = loop {
+            let gap_since_last_used = tweak_idx - last_used_idx.unwrap_or_default();
+
+            args.update_recovery_progress(RecoveryProgress {
+                complete: u32::try_from(tweak_idx.0).unwrap_or(u32::MAX),
+                total: u32::try_from(tweak_idx.0.saturating_add(MAX_GAP)).unwrap_or(u32::MAX),
+            });
+
+            if MAX_GAP < gap_since_last_used {
```

Comment:
I think this causes an issue for backups with an index larger than the max gap.

ex:
  - `backup.next_tweak_idx == 100`
  - first iteration of loop, `last_used_idx == None`
  - `gap_since_last_used == 100`
  - loop breaks, `new_start_tweak_idx == 0 + 3`

Even though we know our next tweak index from the backup is 100, we'd only check the first 3 indexes and save the next peg-in tweak index as 3, causing us to reuse address and potentially miss claiming peg-ins that used an index greater than the backup.

Reply:
Well spotted.

Changing:

```
            let gap_since_last_used = tweak_idx - last_used_idx.unwrap_or(start_tweak_idx);
```

Reply:
Beautiful 👍 

With that approach, you'll also need to `unwrap_or` when we break. Alternatively, I think you can also init  `last_used_idx = start_tweak_idx`, something like

```rust
        let mut last_used_idx = start_tweak_idx;
        let mut current_tweak_idx = start_tweak_idx;
        let mut tweak_idxes_with_pegins = BTreeSet::new();

        debug!(target: LOG_CLIENT_MODULE_WALLET, max_gap=MAX_GAP, %start_tweak_idx, "Checking history");

        let new_start_tweak_idx = loop {
            let gap_since_last_used = current_tweak_idx - last_used_idx;

            args.update_recovery_progress(RecoveryProgress {
                complete: u32::try_from(current_tweak_idx.0).unwrap_or(u32::MAX),
                total: u32::try_from(current_tweak_idx.0.saturating_add(MAX_GAP))
                    .unwrap_or(u32::MAX),
            });

            if MAX_GAP < gap_since_last_used {
                break last_used_idx.advance(NUM_IDX_SKIP_AFTER_LAST_USED);
            }

            let (script, address, _tweak_key, _operation_id) =
                data.derive_peg_in_script(current_tweak_idx);
            btc_rpc.watch_script_history(&script).await?;
            let history = btc_rpc.get_script_history(&script).await?;

            debug!(target: LOG_CLIENT_MODULE_WALLET, %current_tweak_idx, %address, history_len=history.len(), "Checked address");
            if !history.is_empty() {
                last_used_idx = current_tweak_idx;
                tweak_idxes_with_pegins.insert(current_tweak_idx);
            }
            current_tweak_idx = current_tweak_idx.next();
        };
```

Reply:
@bradleystachurski You're saving my butt. I'm tempted to extract the iteration logic to another function so I could easily unit-test it, but I'm afraid the additional boilerplate and abstraction is not worth it.

I'm going with `last_used_idx = start_tweak_idx;`. It makes more logical sense.

Reply:
Ooh, I do love a pure function with some unit tests 🙂 

I think the tradeoff is worthwhile since the potential risks of a bug are address reuse and missing claims during a recovery, but I'll let you make the call.

Reply:
I'll try it them and show it in another commit.

Reply:
@bradleystachurski done